### PR TITLE
SSD train example with multiple GPUs

### PR DESCRIPTION
**Pull request by Hakuyume**

### CODE REVIEW

File: examples/ssd/README.md [hunk 1]
```diff
@@ -34,6 +34,11 @@ Note that this code requires `cv2` module.
 $ python train.py [--model ssd300|ssd512] [--batchsize <batchsize>] [--gpu <gpu>]
 ```
 
+If you want to use multiple GPUs, use `train_mulit.py`.
```
Comment (fukatani) [inline examples/ssd/README.md]: `train_multi.py` is right.
(Sorry, I am reading this PR because it is interesting though this is WIP.)

Comment (Hakuyume) [reply examples/ssd/README.md]: Thank you for pointing out. I will remove `[WIP]` after I confirm this script can reproduce the same performance.

File: examples/ssd/train_multi.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        mb_locs, mb_confs = self.model(imgs)
+        loc_loss, conf_loss = multibox_loss(
+            mb_locs, mb_confs, gt_mb_locs, gt_mb_labels, self.k)
+        loss = loc_loss * self.alpha + conf_loss
+
+        chainer.reporter.report(
+            {'loss': loss, 'loss/loc': loc_loss, 'loss/conf': conf_loss},
+            self)
+
+        # use the number of positives in all GPUs instead of that in this GPU
```
Comment (yuyu2172) [inline examples/ssd/train_multi.py]: use the total number of positives in all GPUs instead of the number of positives in each GPU

File: examples/ssd/train_multi.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        with chainer.cuda.get_device_from_array(gt_mb_labels):
+            if n_positive[0] > 0:
+                loss *= (gt_mb_labels > 0).sum() / n_positive[0]
+        return loss
+
+
+class Transform(object):
+
```
Comment (yuyu2172) [inline examples/ssd/train_multi.py]: If this class is completely same as the training script for single GPU, there should be a comment saying it.

File: examples/ssd/train_multi.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    for param in train_chain.params():
+        if param.name == 'b':
+            param.update_rule.add_hook(GradientScaling(2))
+        else:
+            param.update_rule.add_hook(WeightDecay(0.0005))
+
+    devices = {'main': args.gpu[0]}
+    devices.update(
+        ('other{:d}'.format(i + 1), gpu) for i, gpu in enumerate(args.gpu[1:]))
+    updater = training.ParallelUpdater(train_iter, optimizer, devices=devices)
```
Comment (yuyu2172) [inline examples/ssd/train_multi.py]: Do we need to make `train_multi.py` to behave the same way as `train.py`?
Does training get unstable if make changes?

Also, we should document some changes made from `train.py`, especially ones involving `n_positive`.

Comment (Hakuyume) [reply examples/ssd/train_multi.py]: > Do we need to make train_multi.py to behave the same way as train.py?
Does training get unstable if make changes?

I think it is bad that the number of GPUs affects the behaviour. We should keep the behaviour unchanged as much as possible. (I think it will converge even if we change the behaviour).
Do you want to change the behaviour according to the number of GPUs?

Comment (yuyu2172) [reply examples/ssd/train_multi.py]: > I think it is bad that the number of GPUs affects the behaviour.

Should we follow this rule for all multi-GPU implementations (e.g. ImageNet)?

> Do you want to change the behaviour according to the number of GPUs?

Well, I am wondering if we should use `ParallelUpdater`. Since this is the only option to implement the current behavior, we need to change the behavior if we use other updaters, such as ChainerMN and `MultiprocessParallelUpdater`.
In my opinion, we should always use the same updater for multi-GPU implementation, so that the community would not get confused about which one to use.

Is `ParallelUpdater` fast enough in this training script?
It can be slow because it does not use NCCL.

Comment (Hakuyume) [reply examples/ssd/train_multi.py]: > Should we follow this rule for all multi-GPU implementations (e.g. ImageNet)?

From my understanding, there is no concrete rule for multi-GPU implementations.
In addition, this implementation is not against to that of classification (softmax cross entropy does not change it's behaviour).

> Well, I am wondering if we should use ParallelUpdater. Since this is the only option to implement the current behavior, we need to change the behavior if we use other updaters, such as ChainerMN and MultiprocessParallelUpdater.

I agree that we should consider the cases of  ChainerMN or MultiprocessParallelUpdater.
I will check the APIs of them.

> Is ParallelUpdater fast enough in this training script?

It is not slow when it is compared with Caffe implementation. The biggest motivation of this implementation is to reduce the memory consumption per GPU (please refer Chainer-jp slack).

Comment (yuyu2172) [reply examples/ssd/train_multi.py]: > From my understanding, there is no concrete rule for multi-GPU implementations.

Yes. I would like to discuss if we are going to have any.

> In addition, this implementation is not against to that of classification (softmax cross entropy does not change it's behaviour).

Yeah, but BN needs to be modified if we want to make them the same.

Comment (Hakuyume) [reply examples/ssd/train_multi.py]: BN in ChainerMN keeps the behaviour, doesn't it?

Comment (yuyu2172) [reply examples/ssd/train_multi.py]: I think `MultiNodeBatchNormalization` is not common in ImageNet training protocols, such as the one in 1-hour paper.
So the behavior would change.

Comment (Hakuyume) [reply examples/ssd/train_multi.py]: I heard that PSPNet requires `MultiNodeBatchNormalization` for good performance. We cannot follow the protocol of ImageNet in that case. We don't need to treat `ImageNet in 1 hour` as a standard of all tasks/models.

Comment (Hakuyume) [reply examples/ssd/train_multi.py]: I think the best way is to follow the original papers as much as possible. Changing the behaviour is inconvenient for academic researches because we/users cannot estimate the amount of side-effect of modification.

File: examples/ssd/train_multi.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+        return loss
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '--model', choices=('ssd300', 'ssd512'), default='ssd300')
+    parser.add_argument('--batchsize', type=int, default=32)
+    parser.add_argument('--test_batchsize', type=int, default=32)
```
Comment (yuyu2172) [inline examples/ssd/train_multi.py]: How about making it `16`.

**Comment by yuyu2172** [comment]
LGTM
